main.py

```diff
@@ -104,22 +106,27 @@ def add_switch_metadata(self, dpid):
         """Add metadata to a switch."""
         metadata = request.get_json()
         try:
-            self.controller.switches[dpid].extend_metadata(metadata)
-            return jsonify("Operation successful"), 201
+            switch = self.controller.switches[dpid]
         except KeyError:
             return jsonify("Switch not found"), 404
 
+        switch.extend_metadata(metadata)
+        self.notify_metadata_changes(switch, 'added')
+        return jsonify("Operation successful"), 201
+
     @rest('v3/switches/<dpid>/metadata/<key>', methods=['DELETE'])
     def delete_switch_metadata(self, dpid, key):
         """Delete metadata from a switch."""
         try:
-            status = self.controller.switches[dpid].remove_metadata(key)
+            switch = self.controller.switches[dpid]
         except KeyError:
             return jsonify("Switch not found"), 404
 
-        if status:
-            return jsonify("Operation successful"), 200
-        return jsonify("Metadata not found"), 404
+        if switch.remove_metadata(key) is False:
+            return jsonify("Metadata not found"), 404
```

Comment:
Instead of 404, another option is to just return success if there's nothing to remove. It depends on what will be the best approach for the user of this API.